structuralcodes/codes/mc2020/_concrete_limit_state_of_cracking.py

```diff
@@ -0,0 +1,1367 @@
+            + 'strain calculation in y direction.'
+        )
+
+    return max(
+        (sigma_s_y - beta_TS_y * sigma_sr_ef_y) / Es,
+        sigma_s_y / Es * (1 - beta_TS_y),
+    )
+
+
+def eps_2(tau_xy: float, Ec: float, theta: float) -> float:
+    """Calculate the principal compressive strain eps_2.
+
+    fib Model Code 2020, eq. (30.5-19)
+
+    Args:
+        tau_xy (float): Shear stress in MPa.
+        Ec (float): Modulus of elasticity of concrete in MPa.
+        theta (float): Angle between the x direction reinforcement and the
+            direction of principal tensile stress in degrees.
+
+    Returns:
+        float: Principal compressive strain ε2.
+
+    Raises:
+        ValueError: If any input is invalid.
+    """
+    if tau_xy < 0 or Ec <= 0 or not (0 <= theta <= 90):
+        raise ValueError(
+            'Invalid input values for principal '
+            + 'compressive strain calculation.'
+        )
+
+    theta_rad = math.radians(theta)
+    sin_theta = math.sin(theta_rad)
+    cos_theta = math.cos(theta_rad)
+
+    return -abs(tau_xy) / (Ec * sin_theta * cos_theta)
+
+
+def sigma_s_x(
+    sigma_x: float, tau_xy: float, rho_s_ef_x: float, theta: float
+) -> float:
+    """Calculate the reinforcing steel stress sigma_s,x at the crack in the x
+    direction.
+
+    fib Model Code 2020, eq. (30.5-21)
+
+    Args:
+        sigma_x (float): Normal stress in x direction in MPa.
+        tau_xy (float): Shear stress in MPa.
+        rho_s_ef_x (float): Effective reinforcement ratio in the y direction.
+        theta (float): Angle between the y direction reinforcement and the
+            direction of principal tensile stress in degrees.
+
+    Returns:
+        float: Steel stress sigma_s,y in the x direction in MPa.
+
+    Raises:
+        ValueError: If any input is invalid.
+    """
+    if sigma_x < 0 or tau_xy < 0 or rho_s_ef_x <= 0 or not (0 <= theta <= 90):
+        raise ValueError(
+            'Invalid input values for steel stress calculation in y direction.'
+        )
+
+    theta_rad = math.radians(theta)
+    return (sigma_x + tau_xy * math.tan(theta_rad)) / rho_s_ef_x
+
+
+def sigma_s_y(
+    sigma_y: float, tau_xy: float, rho_s_ef_y: float, theta: float
+) -> float:
+    """Calculate the reinforcing steel stress sigma_s,y at the crack in the y
+    direction.
+
+    fib Model Code 2020, eq. (30.5-21)
+
+    Args:
+        sigma_y (float): Normal stress in y direction in MPa.
+        tau_xy (float): Shear stress in MPa.
+        rho_s_ef_y (float): Effective reinforcement ratio in the y direction.
+        theta (float): Angle between the y direction reinforcement and the
+            direction of principal tensile stress in degrees.
+
+    Returns:
+        float: Steel stress sigma_s,y in the y direction in MPa.
+
+    Raises:
+        ValueError: If any input is invalid
+    """
+    if sigma_y < 0 or tau_xy < 0 or rho_s_ef_y <= 0 or not (0 <= theta <= 90):
```

Comment:
Same comment as before: Should all stresses be strictly positive? Can one or more be negative or zero?